sdk/core/azure-core/src/main/java/com/azure/core/util/Contexts.java

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import java.util.Objects;
+
+/**
+ * A utility type that can be used to add and retrieve instances commonly used in {@link Context}.
+ */
+public final class Contexts {
+
+    private Context context;
+
+    private Contexts(Context context) {
+        this.context = Objects.requireNonNull(context, "'context' must not be null");
+    }
+
+    /**
+     * Creates {@link Contexts} from empty {@link Context}.
+     * @return The {@link Contexts} instance.
+     */
+    public static Contexts empty() {
+        return with(Context.NONE);
+    }
+
+    /**
+     * Creates {@link Contexts} from supplied {@link Context}.
+     * @param context Existing {@link Context}. Must not be null.
+     * @return The {@link Contexts} instance.
+     * @throws NullPointerException If {@code context} is null.
+     */
+    public static Contexts with(Context context) {
+        return new Contexts(context);
+    }
+
+    /**
+     * Adds request's {@link ProgressReporter} instance to the {@link Context}.
+     * @param progressReporter The {@link ProgressReporter} instance.
+     * @return Itself.
+     */
+    public Contexts setHttpRequestProgressReporter(ProgressReporter progressReporter) {
+        context = context.addData(Keys.HTTP_REQUEST_PROGRESS_REPORTER, progressReporter);
+        return this;
+    }
+
+    /**
+     * Retrieves request's {@link ProgressReporter} from the {@link Context}.
+     * @return The {@link ProgressReporter}.
+     */
+    public ProgressReporter getHttpRequestProgressReporter() {
+        return (ProgressReporter) context.getData(Keys.HTTP_REQUEST_PROGRESS_REPORTER).orElse(null);
+    }
+
+    /**
+     * Returns a version of the {@link Context} reflecting mutations.
+     * @return The version of the {@link Context} reflecting mutations.
+     */
+    public Context getContext() {
+        return context;
+    }
+
+    /**
+     * Defines {@link Context} keys commonly used in Azure SDKs.
+     */
+    public static final class Keys {
+        private Keys() {
+        }
+
+        /**
+         * A {@link Context} key for the outgoing request's {@link ProgressReporter}.
+         */
+        public static final String HTTP_REQUEST_PROGRESS_REPORTER = "com.azure.core.http.request.progress.reporter";
+    }
```

Comment:
Does this need to be public? If so, I'd be inclined to move this to Contexts as static fields named KEY_* or *_KEY.

Reply:
For the first question. This is to make insertions in async flows more memory efficient.
I.e. this
```
       ProgressReporter progressReporter = null;
        Flux.just(1)
            .contextWrite(
                ctx -> ctx.put(Contexts.Keys.HTTP_REQUEST_PROGRESS_REPORTER, progressReporter)
            ).subscribe();
```

is less involved than
```
        ProgressReporter progressReporter = null;
        Flux.just(1)
            .contextWrite(
                FluxUtil.toReactorContext(
                    Contexts.empty().setHttpRequestProgressReporter(progressReporter).getContext()
                )
            ).subscribe();
```

For the second question.
We could do `KEY_*` (i.e. prefix is better). However, if I type `Contexts.` in the IDE I'm going to have all those keys show up in hints. Isn't this a concern?